plugins/inputs/metric_streams_listener/metric_streams_listener.go

```diff
@@ -0,0 +1,416 @@
+//go:embed sample.conf
+var sampleConfig string
+
+// defaultMaxBodySize is the default maximum request body size, in bytes.
+// if the request body is over this size, we will return an HTTP 413 error.
+// 500 MB
+const defaultMaxBodySize = 500 * 1024 * 1024
+
+type MetricStreamsListener struct {
+	ServiceAddress string          `toml:"service_address"`
+	Paths          []string        `toml:"paths"`
+	MaxBodySize    config.Size     `toml:"max_body_size"`
+	ReadTimeout    config.Duration `toml:"read_timeout"`
+	WriteTimeout   config.Duration `toml:"write_timeout"`
+	AccessKey      string          `toml:"access_key"`
+
+	requestsReceived selfstat.Stat
+	writesServed     selfstat.Stat
+	requestTime      selfstat.Stat
+	ageMax           selfstat.Stat
+	ageMin           selfstat.Stat
+
+	Log telegraf.Logger
+	tlsint.ServerConfig
+	tlsConf  *tls.Config
+	wg       sync.WaitGroup
+	close    chan struct{}
+	listener net.Listener
+	acc      telegraf.Accumulator
+}
+
+type Request struct {
+	RequestID string `json:"requestId"`
+	Timestamp int64  `json:"timestamp"`
+	Records   []struct {
+		Data string `json:"data"`
+	} `json:"records"`
+}
+
+type Data struct {
+	MetricStreamName string             `json:"metric_stream_name"`
+	AccountID        string             `json:"account_id"`
+	Region           string             `json:"region"`
+	Namespace        string             `json:"namespace"`
+	MetricName       string             `json:"metric_name"`
+	Dimensions       map[string]string  `json:"dimensions"`
+	Timestamp        int64              `json:"timestamp"`
+	Value            map[string]float64 `json:"value"`
+	Unit             string             `json:"unit"`
+}
+
+type Response struct {
+	RequestID string `json:"requestId"`
+	Timestamp int64  `json:"timestamp"`
+}
+
+type age struct {
+	max time.Duration
+	min time.Duration
+}
+
+func (*MetricStreamsListener) SampleConfig() string {
+	return sampleConfig
+}
+
+func (a *age) Record(t time.Duration) {
+	if t > a.max {
+		a.max = t
+	}
+
+	if t < a.min {
+		a.min = t
+	}
+}
+
+func (a *age) SubmitMax(stat selfstat.Stat) {
+	stat.Incr(a.max.Nanoseconds())
+}
+
+func (a *age) SubmitMin(stat selfstat.Stat) {
+	stat.Incr(a.min.Nanoseconds())
+}
+
+func (ms *MetricStreamsListener) Description() string {
+	return "HTTP listener & parser for AWS Metric Streams"
+}
+
+func (ms *MetricStreamsListener) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+// Start starts the http listener service.
+func (ms *MetricStreamsListener) Start(acc telegraf.Accumulator) error {
+	ms.acc = acc
+	server := ms.createHTTPServer()
+
+	tlsConf, err := ms.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", ms.ServiceAddress, tlsConf)
+	} else {
+		listener, err = net.Listen("tcp", ms.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+	ms.tlsConf = tlsConf
+	ms.listener = listener
```

Comment:
Would it be possible to eliminate `tlsConf` from the struct?
```suggestion
	server.TLSConfig, err = ms.ServerConfig.TLSConfig()
	if err != nil {
		return err
	}

	if server.TLSConfig != nil {
		ms.listener, err = tls.Listen("tcp", ms.ServiceAddress, server.TLSConfig)
	} else {
		ms.listener, err = net.Listen("tcp", ms.ServiceAddress)
	}
	if err != nil {
		return err
	}
```

Reply:
Yeah nice, I am able to remove this - this change could also be added to the `http_listener_v2` plugin as well then, it seems like it is unnecessary to use as it is first only used to create the `http.Server` but then updated anyway based on the value returned by `ms.ServerConfig.TLSConfig()`.

Reply:
Its known fact that some of the plugins might need a refactoring. This is something that will be part of Telegraf 2.0